cache.go

```diff
@@ -0,0 +1,152 @@
+// cache ID from the cache.
+//
+// The mutex must be locked when accessing either the stmtDBCache or the
+// dbStmtCache.
+type statementCache struct {
+	// stmtDBCache stores sql.Stmt objects addressed via the cache ID of the
+	// sqlair.Statement they built from and the sqlair.DB they are prepared on.
+	stmtDBCache map[uint64]map[uint64]*sql.Stmt
+
+	// dbStmtCache indicates when a sqlair.Statement has been prepared on a particular sqlair.DB.
+	dbStmtCache map[uint64]map[uint64]bool
+
+	// stmtIDCount and dbIDCount used to generate unique cache IDs.
+	stmtIDCount uint64
+	dbIDCount   uint64
+
+	mutex sync.RWMutex
+}
+
+var once sync.Once
+var singleStmtCache *statementCache
+
+// newStatementCache returns the single instance of the statement cache.
+func newStatementCache() *statementCache {
+	once.Do(func() {
+		singleStmtCache = &statementCache{
+			stmtDBCache: map[uint64]map[uint64]*sql.Stmt{},
+			dbStmtCache: map[uint64]map[uint64]bool{},
+		}
+	})
+	return singleStmtCache
+}
+
+// newStatement returns a new sqlair.Statement and adds it to the cache. A
+// finalizer is set on the sqlair.Statement to remove its ID from the cache and
+// close all associated sql.Stmt objects.
+func (sc *statementCache) newStatement(te *expr.TypedExpr) *Statement {
+	cacheID := atomic.AddUint64(&sc.stmtIDCount, 1)
+	sc.mutex.Lock()
+	sc.stmtDBCache[cacheID] = map[uint64]*sql.Stmt{}
+	sc.mutex.Unlock()
+	s := &Statement{te: te, cacheID: cacheID}
+	// This finalizer is run after the Statement is garbage collected.
+	runtime.SetFinalizer(s, sc.getStmtFinalizer(s))
+	return s
+}
+
+// newDB returns a new sqlair.DB and allocates the necessary resources in the
+// statementCache. A finalizer is set on the sqlair.DB to remove references to
+// it from the cache, close all sql.Stmt objects on it, and close the sql.DB
+func (sc *statementCache) newDB(sqldb *sql.DB) *DB {
+	cacheID := atomic.AddUint64(&sc.dbIDCount, 1)
+	sc.mutex.Lock()
+	sc.dbStmtCache[cacheID] = map[uint64]bool{}
+	sc.mutex.Unlock()
+	db := &DB{sqldb: sqldb, cacheID: cacheID}
+	// This finalizer is run after the DB is garbage collected.
+	runtime.SetFinalizer(db, sc.getDBFinalizer(db))
+	return db
+}
+
+// prepareSubstrate is an object that queries can be prepared on, e.g. a sql.DB
+// or sql.Conn. It is used in prepareStmt.
+type prepareSubstrate interface {
+	PrepareContext(context.Context, string) (*sql.Stmt, error)
+}
+
+// prepareStmt prepares a Statement on a prepareSubstrate. It first checks in
+// the cache to see if it has already been prepared on the DB.
+// The prepareSubstrate must be associated with the same DB that prepareStmt is
+// a method of.
+func (sc *statementCache) prepareStmt(ctx context.Context, dbID uint64, ps prepareSubstrate, s *Statement) (*sql.Stmt, error) {
+	var err error
+	sc.mutex.RLock()
+	// The Statement cache ID is only removed from stmtDBCache when the
+	// finalizer is run. The cache ID must be present as we have a reference to
+	// the Statement.
```

Comment:
I dont understand this sentence, the bit about the reference to the Statement.

Reply:
I've changed this comment to clarify the meaning.

Reply:
It is clear now, thanks